parser.grace

```diff
@@ -1903,16 +1903,12 @@ method typeArgs {
 
 method typeArg {
     // Parses a single type argument, and leaves it on the values stack.
-    // TODO: 'identifier' could be a dotted identifier, 
-    //        or perhaps a type expression?
 
-    if (sym.isIdentifier) then {
-        identifier
-        if (sym.isLGeneric) then {
-            values.push(ast.genericNode.new(values.pop, typeArgs))
-        }
-    } else {
+    // Special case for interface Literals without leading 'interface' keyword.
+    if (sym.isLBrace) then {
         interfaceLiteral
```

Comment:
As far as I recall, the _only_ place that an interface literal is allowed without it's leading `interface` reserved word is as the first thing on the rhs of a type declaration, that is, after 
```
type identifier = 
```
Any place else, we must say `interface`.  So I think that all you really need do is to define typeArg as `expression(noBlocks)`.

A more complete fix would be to actually follow the grammar for typeExpressions, which would exclude `1 + 2` (but not `n + m`, where `n = 1` and `m = 2`).

Reply:
See my new commit, specifically:
* I have remove the special handling of `interface` literals (I only had that their since you're code had it).
* foo[[1 + 2]] is now a syntax error
* Type-Expressions are now also supported in the generics of type-expressions, e.g:
```
method bar(x: Sequence[[String|Number]]) { }
```

Reply:
Thanks.  I'll look at the PR.

> (I only had that there since your code had it)

I'm not clear why you are saying this; in the existing code, interface literals need the leading `interface` keyword everywhere except after `type <name> =`.   I think that we should eliminate that exception; you have proved how confusing it is.

Reply:
>  I think that we should eliminate that exception; you have proved how confusing it is.

I thought we agreed that long ago, and this was still in amg as a matter of backwards compatibility only.

Reply:
>> (I only had that there since your code had it)

> I'm not clear why you are saying this; 

Because you're code was, and I qoute:
```
    if (sym.isIdentifier) then {	
        identifier	
        if (sym.isLGeneric) then {	
            values.push(ast.genericNode.new(values.pop, typeArgs))	
        }	
    } else {	
        interfaceLiteral	
    }
```

Thus if you write `method foo[[T]] {}; foo[[{bar -> Number}]]` on the master branch, you will *not* get a syntax error; whereas on my fork you do.

Regardless, I personally think it's good to minimise the number of special cases in the syntax, specifically when they get confusing:
* `type T = { bar }`, but `type T = A & {bar}` is a syntax error
* `print "hello"`,  but `print hello` is a syntax error.
* `outer.outer`, but `(outer).outer` is a syntax error.

Reply:
Well, actually, I'm a human, and like to be treated that way.  

Yes, you are right!  High time we got rid of that exception anyway.  Having the parser method `interfaceLiteral` do that is just too confusing.

Reply:
Sorry about that @apblack, I did not mean to offend, I was just unsure what the intended behaviour was.